src/neo4j_graphrag/retrievers/cypher.py

```diff
@@ -0,0 +1,379 @@
+#  #
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+from __future__ import annotations
+
+import logging
+import re
+from typing import Any, Callable, Dict, Optional
+
+import neo4j
+from pydantic_core import ErrorDetails
+from neo4j.exceptions import CypherSyntaxError
+from pydantic import ValidationError
+
+from neo4j_graphrag.exceptions import (
+    RetrieverInitializationError,
+    SearchValidationError,
+)
+from neo4j_graphrag.retrievers.base import Retriever
+from neo4j_graphrag.types import (
+    CypherParameterDefinition,
+    CypherParameterType,
+    CypherRetrieverModel,
+    CypherSearchModel,
+    Neo4jDriverModel,
+    RawSearchResult,
+    RetrieverResultItem,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class CypherRetriever(Retriever):
+    """
+    Allows for the retrieval of records from a Neo4j database using a parameterized Cypher query.
+
+    This retriever enables direct execution of predefined Cypher queries with dynamic parameters.
+    It ensures type safety through parameter validation and provides the standard retriever result format.
+
+    Example:
+
+    .. code-block:: python
+
+        import neo4j
+        from neo4j_graphrag.retrievers import CypherRetriever
+
+        driver = neo4j.GraphDatabase.driver(URI, auth=AUTH)
+
+        # Create a retriever for finding movies by title
+        retriever = CypherRetriever(
+            driver=driver,
+            query="MATCH (m:Movie {title: $movie_title}) RETURN m",
+            parameters={
+                "movie_title": {
+                    "type": "string",
+                    "description": "Title of a movie"
+                }
+            }
+        )
+
+        # Use the retriever with specific parameter values
+        results = retriever.search(parameters={"movie_title": "The Matrix"})
+
+    Args:
+        driver (neo4j.Driver): The Neo4j Python driver.
+        query (str): Cypher query with parameter placeholders.
+        parameters (Dict[str, Dict]): Parameter definitions with types and descriptions.
+            Each parameter should have a 'type' and 'description' field.
+            Supported types: 'string', 'number', 'integer', 'boolean', 'array'.
+        result_formatter (Optional[Callable[[neo4j.Record], RetrieverResultItem]]):
+            Custom function to transform a neo4j.Record to a RetrieverResultItem.
+        neo4j_database (Optional[str]): The name of the Neo4j database to use.
+
+    Raises:
+        RetrieverInitializationError: If validation of the input arguments fail.
+    """
+
+    def __init__(
+        self,
+        driver: neo4j.Driver,
+        query: str,
+        parameters: Dict[str, Dict[str, Any]],
```

Comment:
Maybe we can use `list[CypherParameterDefinition]` with the `@validate_call` decorator (see for instance [here](https://github.com/neo4j/neo4j-graphrag-python/blob/main/src/neo4j_graphrag/experimental/components/text_splitters/fixed_size_splitter.py#L94))